### ISSUE TITLE
JIT: Implement or eliminate GTF_VAR_BIRTH

### ISSUE THREAD

**Issue by CarolEidt** [dotnet/coreclr#18196]
This flag is no longer used after the removal of `LEGACY_BACKEND`. The comment above it says that "We should consider using it more generally for VAR_BIRTH, instead of GTF_VAR_DEF && !GTF_VAR_USEASG". This is currently used by `TreeLifeUpdater` (in `UpdateLifeVar()`), and is used in its negation in `fgValueNumberTree()`.
We should either:
- Implement `GTF_VAR_BIRTH` in RyuJIT, ensuring that liveness sets it appropriately, OR
- Eliminate it.

In either case, it would be useful to have properties or methods on `GenTree` that abstract the usage of the liveness flags. This would also entail deciding/agreeing on the terminology - e.g. I would prefer to have `IsPureDef`, `IsLastUse`, `IsDef` and `IsUseDef` over using "birth" and "death" terminology.

**Comment by CarolEidt** [comment]
@dotnet/jit-contrib - this could be an "easy", "up-for-grabs" and/or "Hackathon" issue if the two questions were answered:
- Should we implement it? (Personally I don't think it would buy anything, as testing for one flag set and one flag unset is a simple test)
- What terminology should we use to express these liveness changes: PureDef, LastUse, etc. vs. Birth, Death?

**Comment by mikedn** [comment]
Isn't it `GTF_REG_BIRTH`? I don't see a `GTF_VAR_BIRTH` flag.

> Should we implement it?

I don't understand what's this useful for. If the same information is already provided by other 2 flags then I think there's no doubt that this flag should just be removed. There are already too many flags and some of them exists only due to ancient/legacy design.

**Comment by CarolEidt** [comment]
@mikedn - thanks for the correction (fixed the title)

**Comment by mikedn** [comment]
> This is currently used by TreeLifeUpdater (in UpdateLifeVar()), and is used in its negation in fgValueNumberTree().

It's not clear how useful the stuff in `UpdateLifeVar` is. For example:
* `isBorn` results in `genUpdateVarReg` being called. What would happen if we call `genUpdateVarReg` for every def?
* Both `isBorn` and `isDying` are used to call `genUpdateRegLife` that then calls into `RegSet`. AFAIR a lot of `RegSet` code is useless or debug only code.
* Another use of `isDying` is associated with `indirAddrLocal`. That's dubious, it's not clear why we have this indirs to begin with.
And then the use in `fgValueNumberTree` is IMO not related to "birth" in any way, it just looks for a "use" and it's a side-effect of the pesky `GT_ASG` and its associated processing.

In summary, it's probably not worth bothering too much with new names because there's anyway enough weirdness that won't go away too soon. I'd do:
* `GTF_VAR_USEASG` -> `GTF_VAR_USEDEF` to remove the needless association with ASG inherited from ASG_op. But before doing such a rename I'd check if it's actually needed - you'd think that this is simply implied by `LCL_FLD` with `GTF_VAR_DEF`...
* `GTF_VAR_DEATH` -> `GTF_VAR_LAST_USE` because there are many references to "last use". Though the `death` term seems better in the context of liveness...

**Comment by CarolEidt** [comment]
> But before doing such a rename I'd check if it's actually needed - you'd think that this is simply implied by LCL_FLD with GTF_VAR_DEF...

Actually, a `LCL_FLD` assignment (`GT_ASG(GT_LCL_FLD)` or `GT_STORE_LCL_FLD`) can be used when it is a full definition but with a different type. While I'd like to see that changed to remove that ambiguity, for now the inference you suggest would not be correct.

**Comment by mikedn** [comment]
> Actually, a LCL_FLD assignment (GT_ASG(GT_LCL_FLD) or GT_STORE_LCL_FLD) can be used when it is a full definition but with a different type.

Yes, and if the 2 types have the same size then it's a full definition and if the `LCL_FLD` type is smaller then it is a partial definition. Or, use `BITCAST` when the 2 types are different and have the same size. This should be a rare case so the increase in IR size shouldn't be relevant.

**Comment by CarolEidt** [comment]
> Or, use BITCAST when the 2 types are different and have the same size.

I'm not sure that the `GT_BITCAST` is required prior to `Lowering`. I propose here: https://github.com/dotnet/coreclr/blob/master/Documentation/design-docs/struct-abi.md#high-level-proposed-design that we defer the insertion of the `GT_BITCAST` to `Lowering`, but if that causes type issues in the front-end, we can go ahead and insert them in morph.

**Comment by sdmaclea** [comment]
Morph felt more appropriate to me.

**Comment by mikedn** [comment]
> I'm not sure that the GT_BITCAST is required prior to Lowering

Yes, currently it's only used during lowering but I don't see anything about it that would prevent it from being used earlier. In fact, I once did that experimentally in morph to deal with something like `*((int*)&float_var)` and it worked fine.
